src/repeater.rs

```diff
@@ -0,0 +1,258 @@
+        for child in &mut self.children {
+            let lens = &child.lens;
+            let widget = &mut child.widget;
+            self.list_lens.with_mut(data, |data| {
+                lens.with_mut(data, |data| widget.event(ctx, event, data, env))
+            });
+        }
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        if let LifeCycle::WidgetAdded = event {
+            let children = &mut self.children;
+            let child_generator = &self.child_generator;
+            let id_getter = &self.id_getter;
+            self.list_lens.with(data, |data| {
+                for i in 0..data.len() {
+                    let item = &data[i];
+                    children.push(RepeaterChild {
+                        widget: WidgetPod::new(Box::new((child_generator)(item))),
+                        id: (id_getter)(item),
+                        lens: ListItemLens::<U>::new(i),
+                    });
+                }
+            });
+            ctx.children_changed();
+        }
+        for child in &mut self.children {
+            let lens = &child.lens;
+            let widget = &mut child.widget;
+            self.list_lens.with(data, |data| {
+                lens.with(data, |data| widget.lifecycle(ctx, event, data, env))
+            });
+        }
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &T, data: &T, env: &Env) {
+        let list_lens = &self.list_lens;
+        let id_getter = &self.id_getter;
+        let children = &mut self.children;
+        let child_generator = &self.child_generator;
+
+        let mut will_diff = false;
+
+        list_lens.with(old_data, |old_list| {
+            list_lens.with(data, |list| {
+                // Before we start, check if there are any differences
+
+                if old_list.len() != list.len() {
+                    will_diff = true;
+                } else {
+                    // If the lists have the same length, loop through and
+                    // check equality on each ID
+                    for i in 0..list.len() {
+                        if !(id_getter)(old_list.get(i).unwrap())
+                            .same(&(id_getter)(list.get(i).unwrap()))
+                        {
+                            will_diff = true;
+                            break;
+                        }
+                    }
+                }
+
+                if !will_diff {
+                    return;
+                }
+
+                // Diffing has three stages: remove, add and sort. For each, it
+                // figures out what of each operation must be performed to
+                // transform old_list into list, and performs those on the
+                // widget list.
+
+                // If we didn't add or remove anything, this will remain false.
+                let mut children_changed = false;
+
+                // Start
+                let mut stale = HashSet::new();
```

Comment:
I'd probably have done it that way at first. And maybe what follows is moot because I'm not familiar enough with Rust internals. But assuming C++ knowledge transposes somewhat, I'd say it's unlikely that you will have such a large collection that a linear search into a `Vec` becomes a bottleneck. And so instead of using a complex `HashSet`-based add/remove, I'd simply compare 2 sorted vectors, which can be done in linear time. Most likely for moderate-sized collections this will be both simpler and faster.

Reply:
...and if you can use vectors, then it sounds to me like you can use [`itertools.merge_join_by()`](https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.merge_join_by) to greatly reduce the work on iterating, and just focus on the handling of added/removed items.

Reply:
You're likely right that the simplest solution is also the fastest in most cases. It could make sense to branch by collection size here? I don't want this to ever be a bottleneck since `update` is called a lot, but I also want to avoid unnecessary performance overhead in the average case.

> I'd simply compare 2 sorted vectors

The inputs are not sorted, but I do have to preserve their order. The correct ordering of the widgets is determined by the position of each item in the vector from `data`. Since `data` represents the most current state, my goal is to get the ordering of my widgets to match that ordering, so in a sense the only valid sorting order is the order that the vector from `data` is already in. But the order of the vector from `old_data` is important too, as it should always match the current order of my widgets when the algorithm starts. So in a sense my goal is to find what operations need to be applied to `old_data`'s vector to turn it into `data`'s vector, and then apply those operations to my widget list.

Aside from this pre-existing ordering, I have nothing to sort by. Each data item has an ID that can be fetched using the `self.id_getter` closure. This ID must be hashable but not necessarily orderable. Maybe it's not too restricting to add `Ord` as a constraint? But then I still have to allocate memory for two new arrays and eat some CPU and memory cycles to sort them (since I still need the original ordering), and at that point I can't imagine there are performance wins at any collection size since both methods eat memory. The advantage with the current method is that, by creating a HashMap with `<ID, target index>`, you gain O(1) lookup of the target index for each item. I need to perform this lookup for the sort at the end. If that lookup becomes O(n), then the algorithm has `O(log(n)*n^2)` worst-case complexity, whereas it currently has worst-case `O(n*log(n))`.

As I said above, I think it may still be possible to squeeze out some performance using a somewhat more in-place method, but this is speculative and I do want to profile some before I commit to implementing it.